src/Simulation/Common/StackTrace.cs

```diff
@@ -135,36 +136,94 @@ public static StackFrame[] PopulateSourceLocations(Stack<StackFrame> qsharpCallS
+            while (csharpFrame != null && qsharpFrame != null)
             {
-                string fileName = csharpCallStack[csharpStackFrameId].GetFileName();
-                if (fileName != null)
+                if (IsMatch(csharpFrame, qsharpFrame))
                 {
-                    fileName = System.IO.Path.GetFullPath(fileName);
-                    int failedLineNumber = csharpCallStack[csharpStackFrameId].GetFileLineNumber();
-                    StackFrame currentQsharpStackFrame = qsharpStackFrames[qsharpStackFrameId];
-                    if (fileName == currentQsharpStackFrame.SourceFile &&
-                        currentQsharpStackFrame.DeclarationStartLineNumber <= failedLineNumber &&
-                            (
-                                (failedLineNumber < currentQsharpStackFrame.DeclarationEndLineNumber) ||
-                                (currentQsharpStackFrame.DeclarationEndLineNumber == -1)
-                            )
-                        )
-                    {
-                        currentQsharpStackFrame.FailedLineNumber = failedLineNumber;
-                        qsharpStackFrameId++;
-                        if (qsharpStackFrameId == qsharpStackFrames.Length) break;
-                    }
+                    PopulateQSharpFrameFromCSharpFrame(csharpFrame, qsharpFrame);
+                    // Advance to the next Q# Stack frame only when it matches C# stack frame.
+                    // This assumes all Q# stack frames with enough information to match will match with corresponding C# stack frames.
+                    // If corresponding C# stack frame doesn't have enough information to match
+                    // this and all subsequent Q# stack frames will not match.
+                    qsharpFrame = GetNextQSharpStackFrame(qsharpStackFrames, ref currentQSharpFrameIndex);
                 }
+                csharpFrame = GetNextCSharpStackFrame(csharpStackFrames, ref currentCSharpFrameIndex);
             }
             return qsharpStackFrames;
         }
+
+        /// <summary>
+        /// Return next Q# stack frame that has enough information to match
+        /// </summary>
+        private static StackFrame GetNextQSharpStackFrame(StackFrame[] qsharpStackFrames, ref int currentFrameIndex)
+        {
+            while (currentFrameIndex < qsharpStackFrames.Length)
+            {
+                StackFrame currentFrame = qsharpStackFrames[currentFrameIndex];
+                currentFrameIndex++;
+                if (!string.IsNullOrEmpty(currentFrame.SourceFile))
+                {
+                    return currentFrame;
+                }
+            }
+            return null;
+        }
+
+        /// <summary>
+        /// Return next C# stack frame that has enough information to match
+        /// </summary>
+        private static System.Diagnostics.StackFrame GetNextCSharpStackFrame(System.Diagnostics.StackFrame[] csharpStackFrames, ref int currentFrameIndex)
```

Comment:
>GetNextCSharpStackFrame [](start = 53, length = 23)

Let's simplify to currentFrameIndex++; without skipping empties.

Reply:
Updated code not to skip C# frames.